### ISSUE TITLE
Missing keyboard shortcut features from Sublime Text

### ISSUE THREAD

**Issue by waderyan** [Microsoft/vscode#3776]
I created an [extension](https://marketplace.visualstudio.com/items?itemName=ms-vscode.sublime-keybindings) to port Sublime keybindings to VS Code keybindings. I made a list of all the key bindings that were missing analogous commands in VS Code. 

Here is the list roughly ordered by importance. Please let me know about more I missed in the comments. If an extension exists with the functionality I added it in the far right column. 

| Windows / Linux | Mac | Description | Issue | Extension |
| --- | --- | --- | --- | --- |
| <kbd>ctrl</kbd>+left mouse click | <kbd>cmd</kbd>+left mouse click | Add multi cursors | https://github.com/Microsoft/vscode/issues/3130 | None | 
| <kbd>ctrl</kbd>+<kbd>;</kbd> | None | Go to Word | https://github.com/Microsoft/vscode/issues/14028 | None |
| <kbd>ctrl</kbd>+<kbd>+</kbd> | <kbd>cmd</kbd>+<kbd>+</kbd> and <kbd>cmd</kbd>+<kbd>=</kbd> | Increase Font Size | https://github.com/Microsoft/vscode/issues/14031 | [FontSize Shortcuts](https://marketplace.visualstudio.com/items?itemName=peterjuras.fontsize-shortcuts) |
| <kbd>ctrl</kbd>+<kbd>-</kbd> | <kbd>cmd</kbd>+<kbd>-</kbd> | Decrease Font Size | https://github.com/Microsoft/vscode/issues/14031 | [FontSize Shortcuts](https://marketplace.visualstudio.com/items?itemName=peterjuras.fontsize-shortcuts) |
| <kbd>ctrl</kbd>+<kbd>shift</kbd>+<kbd>m</kbd> | <kbd>ctrl</kbd>+<kbd>shift</kbd>+<kbd>m</kbd> | Expand selection to brackets | https://github.com/Microsoft/vscode/issues/14089 | [Bracket Selection](https://marketplace.visualstudio.com/items?itemName=guosong.bracketselection) |
| <kbd>ctrl</kbd>+<kbd>shift</kbd>+<kbd>j</kbd> | <kbd>cmd</kbd>+<kbd>shift</kbd>+<kbd>j</kbd> | Expand selection to indent level | https://github.com/Microsoft/vscode/issues/14090 | None |
| <kbd>alt</kbd>+<kbd>q</kbd> | <kbd>cmd</kbd>+<kbd>alt</kbd>+<kbd>q</kbd> | Break / Wrap line | https://github.com/Microsoft/vscode/issues/14093 | None |
| <kbd>alt</kbd>+<kbd>shift</kbd>+<kbd>1</kbd> | <kbd>alt</kbd>+<kbd>cmd</kbd>+<kbd>1</kbd> | Switch to single column view | https://github.com/Microsoft/vscode/issues/14094 | None |
| <kbd>ctrl</kbd>+<kbd>y</kbd> | <kbd>ctrl</kbd>+<kbd>y</kbd> | Repeat | https://github.com/Microsoft/vscode/issues/14097 | None |
| <kbd>ctrl</kbd>+<kbd>shift</kbd>+<kbd>a</kbd> | <kbd>cmd</kbd>+<kbd>shift</kbd>+<kbd>a</kbd> | Expand selection to tag | https://github.com/Microsoft/vscode/issues/14098 | [expand-region](https://marketplace.visualstudio.com/items?itemName=letrieu.expand-region) |
| <kbd>ctrl</kbd>+<kbd>shift</kbd>+<kbd>space</kbd> | <kbd>cmd</kbd>+<kbd>shift</kbd>+<kbd>space</kbd> | Expand selection to scope | https://github.com/Microsoft/vscode/issues/14099 | [expand-region](https://marketplace.visualstudio.com/items?itemName=letrieu.expand-region) |
| <kbd>shift</kbd>+<kbd>f11</kbd> | <kbd>ctrl</kbd>+<kbd>shift</kbd>+<kbd>cmd</kbd>+<kbd>f</kbd> | Enter distraction free mode | https://github.com/Microsoft/vscode/issues/12940 | None |
| <kbd>ctrl</kbd>+<kbd>space</kbd> | <kbd>ctrl</kbd>+<kbd>space</kbd> | Select next auto complete suggestion | https://github.com/Microsoft/vscode/issues/14101 | None |
| <kbd>alt</kbd>+<kbd>shift</kbd>+<kbd>w</kbd> | <kbd>ctrl</kbd>+<kbd>shift</kbd>+<kbd>w</kbd> | Wrap HTML tag | https://github.com/Microsoft/vscode/issues/14103 | None |
| <kbd>alt</kbd>+<kbd>.</kbd> | <kbd>cmd</kbd>+<kbd>alt</kbd>+<kbd>.</kbd> | Close HTML tag | https://github.com/Microsoft/vscode/issues/14104 | [Close HTML/XML tag](https://marketplace.visualstudio.com/items?itemName=Compulim.compulim-vscode-closetag) |
| <kbd>ctrl</kbd>+<kbd>k</kbd> <kbd>ctrl</kbd>+<kbd>6</kbd> | <kbd>cmd</kbd>+<kbd>k</kbd> <kbd>cmd</kbd>+<kbd>6</kbd> | Fold level 6 | https://github.com/Microsoft/vscode/issues/14106 | None |
| <kbd>ctrl</kbd>+<kbd>k</kbd> <kbd>ctrl</kbd>+<kbd>7</kbd> | <kbd>cmd</kbd>+<kbd>k</kbd> <kbd>cmd</kbd>+<kbd>7</kbd> | Fold level 7 | https://github.com/Microsoft/vscode/issues/14106 | None |
| <kbd>ctrl</kbd>+<kbd>k</kbd> <kbd>ctrl</kbd>+<kbd>8</kbd> | <kbd>cmd</kbd>+<kbd>k</kbd> <kbd>cmd</kbd>+<kbd>8</kbd> | Fold level 8 | https://github.com/Microsoft/vscode/issues/14106 | None |
| <kbd>ctrl</kbd>+<kbd>k</kbd> <kbd>ctrl</kbd>+<kbd>9</kbd> | <kbd>cmd</kbd>+<kbd>k</kbd> <kbd>cmd</kbd>+<kbd>9</kbd> | Fold level 9 | https://github.com/Microsoft/vscode/issues/14106 | None |
| <kbd>alt</kbd>+<kbd>0</kbd> | <kbd>cmd</kbd>+<kbd>0</kbd> | Select open tab index 9 | https://github.com/Microsoft/vscode/issues/14095 | None |

Additionally, here are other actions that are not bound to a keyboard shortcut:

| Action | Issue | Extension |
| --- | --- | --- |
| Drag and drop selected text | https://github.com/Microsoft/vscode/issues/1046 | None |

* Customize keybinding support in menus (see this comment https://github.com/Microsoft/vscode/issues/3776#issuecomment-260324787)
* Customize keybinding support in viewlets (see this https://github.com/Microsoft/vscode/issues/4557)

**Comment by Tyriar** [comment]
Is <kbd>super</kbd> mapped to <kbd>ctrl</kbd> or <kbd>alt</kbd>? Surely they don't bind <kbd>win</kbd>+<kbd>&lt;num&gt;</kbd>

**Comment by waderyan** [comment]
<kbd>super</kbd> is <kbd>win</kbd> or <kbd>cmd</kbd>

**Comment by Tyriar** [comment]
@waderyan just looked, I don't see any `super` in the default keybindings on Windows. Is the above the OSX version?

**Comment by waderyan** [comment]
Yes. From the [docs](http://sublimetext.info/docs/en/reference/key_bindings.html) it looked like this was the case

**Comment by JamesHenry** [comment]
(Hopefully) new VSCode convert here! One of my most used commands in sublime is `shift` + `cmd` + `l` which sublimes calls 'Split into Lines' (under "Selection" menu).

It creates a new cursor for each line of a given selection. Does this functionality exist in VSCode? I cannot see it in the list of missing bindings @waderyan compiled above, but I also cannot seem to find it in VSCode.

**Comment by waderyan** [comment]
@JamesHenry not sure how I missed that one. Added now.

**Comment by JamesHenry** [comment]
Great, thanks. Consider this my +1 :)

**Comment by drew-wallace** [comment]
`[
    { "keys": ["ctrl+minus"], "command": "jump_back" }, //this command already exists in vscode as cursorUndo
    { "keys": ["ctrl+shift+minus"], "command": "jump_forward" } //this one has yet to be added
]`

**Comment by yisibl** [comment]
@JamesHenry In vscode we can used:

<kbd>cmd</kbd>  + <kbd>i</kbd> and

<kbd>option</kbd> + <kbd>shift</kbd> + <kbd>i</kbd>

**Comment by JamesHenry** [comment]
Ah, awesome thanks @yisibl!

Yeah, I have just made it my first ever shortcut override :D

**Comment by yisibl** [comment]
By the way, VS Code not support <kbd>cmd</kbd> + <kbd>j</kbd>

**Comment by kabirbaidhya** [comment]
It would be awesome if we have all these in vscode :+1:

**Comment by Tyriar** [comment]
@kabirbaidhya you can install the [Sublime Keymap extension](https://marketplace.visualstudio.com/items?itemName=ms-vscode.sublime-keybindings) that @waderyan built right now. If you find any issues you can file them to the repo https://github.com/Microsoft/vscode-sublime-keybindings/issues

**Comment by nnur** [comment]
My #1 most used sublime shortcut is:

Ctrl + ⇧ + ↑	Move line/selection up
Ctrl + ⇧ + ↓	Move line/selection down

Any chance of adding these to the package?

**Comment by Zarel** [comment]
I wrote an implementation of Join Lines and Transpose (and some other Sublime commands) which are more faithful to how Sublime implements them:

https://marketplace.visualstudio.com/items?itemName=Zarel.sublime-commands

I generally don't think it's a big deal to copy another editor's behavior exactly, but in this case, the main difference is that Sublime's behavior is less buggy. Specifically:

- transpose crashes when some selections are empty and others aren't; sublime-commands transposes them as expected
- transpose behaves weirdly when there are three or more selections; sublime-commands transposes them like Sublime Text
- join-lines crashes when a cursor is on the last line, sublime-commands doesn't
- join-lines doesn't join the last line of a multiline selection, sublime-commands does (matching Sublime Text's behavior)

If you agree that these are better, could you update the readme in https://github.com/Microsoft/vscode-sublime-keybindings and this issue to refer to those?

**Comment by Zarel** [comment]
Also: Would you be opposed to pull requests adding these features directly into VS Code itself? These features are presumably broadly useful enough to be built-in to Sublime, and then we wouldn't need to install extensions to get access to them.

**Comment by kevinSuttle** [comment]
@nnur I have been itching for the same thing. Transpose _sort_ of does it, but not the way we're used to.

**Comment by gkalpak** [comment]
:+1: to @Zarel's [sublime-commands][1]. The behavior for joining lines and adding cursors to selected lines are closer to Sublime Text's.

[1]: https://marketplace.visualstudio.com/items?itemName=Zarel.sublime-commands

**Comment by claudio4** [comment]
Please also add the crtl + click to add a new cursor.

BTW great work with the extension!

**Comment by waderyan** [comment]
@claudio4 added!

To all - any thoughts on prioritizing the above table? I think join lines is high, followed by ctrl+click to add new cursor. From my experience, I really miss paste and indent. More feedback is appreciated. We have working through a couple of these on our [iteration plan](https://github.com/Microsoft/vscode/issues/15099) for this month.

**Comment by rebornix** [comment]
@waderyan for #9415, we already have all support.

**Comment by gkalpak** [comment]
FWIW, the two main Sublime features I would terribly miss are:

1. The ability to quickly switch between projects (`Project > Quick Switch Project`)
2. Magically restoring my unsaved files when switching between projects or closing/reopening windows.

<sub>(But, of course, it is are more about missing functionality than just missing shortcuts.)</sub>

**Comment by Omen-X** [comment]
Hi, tell me please, i'm binding moving up and down on shortcuts (like alt+j, alt+k), in sublime i can moving up and down with this hotkeys when suggestion menu is open  (or another menu). Can i realize it in VS ?

**Comment by Tyriar** [comment]
@gkalpak I recommend checking out one of the project manager extensions for that, I use [Git Project Manager](https://marketplace.visualstudio.com/items?itemName=felipecaputo.git-project-manager) personally.

Hot exit is coming in the next version, see #101

**Comment by waderyan** [comment]
@Omen-X that is not currently supported, but I have added it to my list.

**Comment by gkalpak** [comment]
@Tyriar, awesome to see hot-exit making it into VSCode 👍 :dancing_men: 

Git Project Manager is pretty nice actually, but I miss a more flexible "project" abstraction (mainly for multi-directory projects). Fortunately, it is on the roadmap (and I can leave without it for a while :smiley:).

<sub>#AboutToSwitch</sub>

**Comment by Tyriar** [comment]
:+1:, FYI @gkalpak multiple folder workspaces is tracked in https://github.com/Microsoft/vscode/issues/396

**Comment by waderyan** [comment]
FYI there is a PR out to add functionality here - https://github.com/Microsoft/vscode/pull/15787

**Comment by thekalinga** [comment]
Can a shortcut be added to toggle case along with upper and lowercase, as this makes life easier to have just one shortcut to toggle case.

Which is what IDEA follows

**Comment by waderyan** [comment]
@thekalinga that is in the PR mentioned above 👍

**Comment by thekalinga** [comment]
Hi,

I see only these two

Upper case	https://github.com/Microsoft/vscode/issues/14096	change case
Lower case	https://github.com/Microsoft/vscode/issues/14096

What I was asking is to have an action that can toggle. i.e if the shortcut is `ctrl+t`, this toggles the current case to uppercase by default, if you press it again, the selected text would be converted to lowercase.

Can u please confirm

Thanks

**Comment by waderyan** [comment]
@thekalinga thanks for the clarification. Can you open a separate issue for that? This issue is specific to keyboard shortcuts in Sublime Text. Sublime Text has separate commands for upper and lower case.

**Comment by thekalinga** [comment]
Sure. I'll open one for that.

The reason why I added a comment here is because since upper & lower are getting implemented, it would be easy for contributors to make the change in one shot.

Thanks in anycase

**Comment by waderyan** [comment]
@thekalinga makes sense. Thanks for your feedback!

**Comment by waderyan** [comment]
Note that we closed on some of these issues for VS Code 1.8 release. 

- [x] Join lines
- [x] Transpose
- [x] Upper/Lower case
- [x] Paste and format

**Comment by Zarel** [comment]
Your join lines implementation doesn't work the same way Sublime does when text is selected, and neither does your transpose implementation. :( The join lines difference is minor, but are you open to pull requests changing Transpose?

**Comment by rebornix** [comment]
@Zarel we are definitely open to pull requests. The reason that you are seeing differences is

* Vim, Sublime Text and Atom all have Join Lines, they are not totally the same when dealing with leading/trailing spaces. If you see any other difference, feel free to change it or file an issue.
* Sublime Text 3 and 2 are doing Transpose differently. Here we mimic Sublime Text 3.

If you'd like to file pull requests, please describe what kind of behavior in your new implementation and why you think it's better or more reasonable than the original one. Thanks!

**Comment by Zarel** [comment]
@rebornix There still exists a Sublime Text 3 behavior VS Code doesn't mimic.

Sublime Text 3 will, if some selections aren't empty, transpose the selections, putting the contents of the first selection into the second, the second into the third, etc, and the last into the first.

For instance, if you have `foo bar`, and you select `foo` and `bar` and do a transpose, you will have `bar foo`.

The use-case is stuff like:

    if (key === 'up') goUp();
    if (key === 'down') goDown();

I can doubleclick on `goUp`, alt+doubleclick (or cmd+doubleclick in Sublime) on `goDown`, and then ctrl+t to get:

    if (key === 'up') goDown();
    if (key === 'down') goUp();

In general, being able to easily swap two variable or two literals is surprisingly useful. Without this, the swap is pretty annoying to do. Something like, double-click goUp, cmd+c, double-click goDown, left, cmd+v, shift+alt+right, cmd+c, click where goUp used to be (a much smaller click target than before), cmd+v. I'd often retype them rather than deal with that mess.

That's the behavior my plugin implements, and the behavior I'd like to pullreq.

**Comment by rebornix** [comment]
@Zarel basically I like the idea. I did implement the word transposing at the beginning but since it requires more consideration about multi cursor mode so I didn't rush it in. We do have API for you to detect word boundary so it would be not difficult to add this feature, but like I said, multi cursor is kind of the monster. As you have already have implemented it in your extension, it would be awesome if you can contribute to the core.

**Comment by bpasero** [comment]
> Customize keybinding support in viewlets (see this #4557)

This is added for February.

**Comment by kcdipesh** [comment]
How about ctrl +shift + d  for duplicating the current line?

**Comment by tangmi** [comment]
On Windows, Sublime Text has a very useful "keybinding" to switch editor tabs with `right mouse button`+`scroll wheel up/down`. I wonder how difficult it would be to add support for this?

Seems related to #3130, however there's no mention of pure mouse keybindings.

**Comment by iagowp** [comment]
I'd like to see ctrl + shift + v / cmd + shift + v to paste and ident

**Comment by iagowp** [comment]
I'd like to see this:
Add other column(s) to selection by click & drag: Ctrl + Shift + RightMouseBtn. -> windows hotkey

**Comment by RamyElkest** [comment]
`ctrl`+`shift`+`up`/`down`
https://www.jflh.ca/2016-07-10-move-lines-up-and-down-in-visual-studio-code
 [
   {
     "key": "ctrl+shift+up",
     "command": "editor.action.moveLinesUpAction",
     "when": "editorTextFocus"
   },
   {
     "key": "ctrl+shift+down",
     "command": "editor.action.moveLinesDownAction",
     "when": "editorTextFocus"
   }
 ]

**Comment by elboletaire** [comment]
@RamyElkest those actually exist... they're `Move Line Up` and `Move Line Down`.

**Comment by ssteinerx** [comment]
Ctrl-K/Ctrl-Y for yank on macOS.  

Ctrl-K cuts to end of line
Ctrl-Y yanks it back

This is ideally separate from the normal cut&paste buffer i.e. Ctrl-K followed by Cmd-V will paste the last normally cut/copied text, not the Ctrl-K buffer which can only be retrieved by Ctrl-Y

**Comment by albert-001** [comment]
Shift + Righ Mouse Click, this is very useful for editing multi-lines at the same time.

**Comment by mgiraldo** [comment]
i really like sublime's ⌘+T that shows a Go To File... dialog... you then start typing characters in then dialog and files that contain those characters get listed... like a very fuzzy autocomplete

super useful for heavily-foldered projects (rails, react)

**Comment by cmalard** [comment]
<kbd>ctrl</kbd>+<kbd>shift</kbd>+<kbd>d</kbd> with text selection : `duplicateText`
extension [sublime-duplicate-text](https://marketplace.visualstudio.com/items?itemName=brian-anders.sublime-duplicate-text)

**Comment by elboletaire** [comment]
The "duplicate selection" exists in vscode. It's called `copy line down`.

**Comment by cmalard** [comment]
This is exactly what I'm talking about : duplicate selection, not the whole line ;-)

**Comment by darkvertex** [comment]
Sadly sublime-duplicate-text does only selection and if you have nothing selected, it doesn't do the whole line like Sublime would. :-1: :(

**Comment by djdeo** [comment]
sublime uses `ctrl+shift+;` to delete html tags (both open tag and close tag), what's the shortcut in vscode?

**Comment by joncalhoun** [comment]
Is there a good way to replicate ST's layouts and their keyboard shortcuts in VSCode, even if it means using another plugin? (screenshot below of the options)

![screenshot of st layouts](https://www.dropbox.com/s/alpyw0ftxkqslkv/Screenshot%202018-02-06%2013.30.21.png?dl=0&raw=1)

I can sorta use the split editor keyboard shortcut to constant split editors and go from 1 column to many, and I can use another keyboard shortcut to swap orientation, but there doesn't appear to be any way to create a shortcut like ST's cmd+alt+1 which would merge them all into a single column, which I use quite often.

**Comment by jaegz** [comment]
One of my all time favorite sublime shortcuts which I greatly miss in VS Code...

**Select all inside tags / Expand selection outwards** 
<kbd>ctrl</kbd>+<kbd>,</kbd> or <kbd>ctrl</kbd>+<kbd>shift</kbd>+<kbd>a</kbd>

**Comment by giovannipds** [comment]
What about to <kbd>ctrl</kbd>+<kbd>up</kbd> and <kbd>ctrl</kbd>+<kbd>down</kbd> (arrows) to increment / decrement selected numbers? How to achieve it?

**Comment by giovannipds** [comment]
Just got it: https://marketplace.visualstudio.com/items?itemName=nmsmith89.incrementor (it's necessary to set the keybindings provided in the page and reload).

**Comment by elboletaire** [comment]
@giovannipds that is an emmet function, and it's already on code. No need for additional plugins.

**Comment by giovannipds** [comment]
@elboletaire right, thank you! I must haven't noticed. But the plugin does some additional stuff like swapping `true` and `false` etc., which is also cool. Will let it installed anyway. =D

**Comment by tino** [comment]
I'm missing these!
```
{ "keys": ["super+e"], "command": "slurp_find_string" },
{ "keys": ["super+shift+e"], "command": "slurp_replace_string" },
```

**Comment by giovannipds** [comment]
Is there a way to switch between Workspaces in Code just like SublimeText does with Projects? That's pretty needed... =(

**Comment by janderson099** [comment]
> I'm missing these!
> 
> ```
> { "keys": ["super+e"], "command": "slurp_find_string" },
> { "keys": ["super+shift+e"], "command": "slurp_replace_string" },
> ```

<kbd>super</kbd> key, on VSCode is <kbd>win</kbd> (I'm using win10).
For instance, I did this code on keybinding.json:
```
{
        "key": "shift+win+l",
        "command": "editor.action.insertSnippet",
        "when": "editorTextFocus",
        "args": {
            "langId": "javascript",
            "name": "Print to console"
        }
    }
```

I hope this help someone.

**Comment by andig** [comment]
You mean Cmd-Shift-Right Del?

**Comment by Victor-Bernabe** [comment]
> You mean Cmd-Shift-Right Del?

I mean the possibility to assign that action to a simple shortcut like ctrl + k

**Comment by andig** [comment]
See https://github.com/Microsoft/vscode/issues/25075

**Comment by ddayhtx** [comment]
multi cursor select is NOT working. Before I wiped the drives and installed OS in both my machines,  the Sublime extension, the ONE thing I want and use the most is no longer working. A month ago, it worked. Now, not on either. I went over the posts above. I am obviously missing something. I'm running Mint 19 Tessa on both machines. 

Any thoughts here? I LOVE VSC but without the multi cursor I'm back to Sublime. It's just not the same :(
